compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java

```diff
@@ -1134,6 +1160,106 @@ public void visit(BLangSimpleVariable varNode) {
         env.enclPkg.isListenerAvailable |= Symbols.isFlagOn(varNode.symbol.flags, Flags.LISTENER);
     }
 
+    @Override
+    public void visit(BLangUserDefinedType userDefinedType) {
+        if (Types.getReferredType(userDefinedType.getBType()).getKind() == TypeKind.RECORD) {
+            visitTypedesc(userDefinedType.pos, userDefinedType.getBType(), Collections.emptyList());
+        }
+    }
+
+    @Override
+    public void visit(BLangValueType valueType) {
+    }
+
+    @Override
+    public void visit(BLangUnionTypeNode unionTypeNode) {
+        unionTypeNode.memberTypeNodes.forEach(typeNode -> typeNode.accept(this));
+    }
+
+    @Override
+    public void visit(BLangRecordTypeNode recordTypeNode) {
+        for (BLangSimpleVariable field : recordTypeNode.fields) {
+            if (field.typeNode != null) {
+                field.typeNode.accept(this);
+            }
+        }
+        if (recordTypeNode.restFieldType != null) {
+            recordTypeNode.restFieldType.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(BLangArrayType arrayType) {
+        arrayType.elemtype.accept(this);
+    }
+
+    @Override
+    public void visit(BLangConstrainedType constrainedType) {
+        constrainedType.constraint.accept(this);
+    }
+
+    @Override
+    public void visit(BLangErrorType errorType) {
+        if (errorType.detailType != null) {
+            errorType.detailType.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(BLangFunctionTypeNode functionTypeNode) {
```

Comment:
Don't we have to visit parameter and return types?

Reply:
Updated https://github.com/ballerina-platform/ballerina-lang/pull/39101/commits/ed9d56f60b01a2744651c7e805361dd574b8357d

Reply:
What about

```ballerina
public function main() {
    function (record {| float f; |}) returns record {| string b = "hello"; |} y;
}
```

Where are the new typedesc instructions generated for these records?

Reply:
Didn't generate `typedesc` for this because there is no usage of defining typedescs for these records.

Reply:
What do you mean no usage?

```ballerina
import ballerina/io;

annotation record {| float f = foo(); |} annot on record field;

public function main() {
    function (record {| @annot float f; |}) r;
}

isolated function foo() returns float {
    io:println("foo");
    return 1.0;
}
```

This should print `"foo"`, but it doesn't atm.

Either way, we shouldn't make assumptions about usage. The spec defines when the typedesc is resolved, and we have to resolve it at that point.

Reply:
Updated.

Reply:
I assume we need changes in https://github.com/ballerina-platform/ballerina-lang/pull/38488 for the sample I shared above to work as expected? We need to make sure we have tests for every possible type node scenario.

Reply:
Did we add this as a test in the other PR?

Reply:
@chiranSachintha, any update on this? Doesn't seem to work for me with the other PR also, will recheck.

Reply:
Can you send a draft PR with changes from both PRs, resolving conflicts?